### PR TITLE
`multi-arch-builder-controller`: Better logging and tests

### PR DESCRIPTION
**Pull request by danilo-gemoli**
I had hard time trying to understand what was going wrong with a failing build and I have realized we could have done much better with logs.

/cc @droslean

### CODE REVIEW

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go [hunk 1]
```diff
@@ -394,6 +430,13 @@ func TestReconcile(t *testing.T) {
 					},
 				},
 			},
+			wantLog: logOutput(func(e *logrus.Entry) {
```
Comment (droslean) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:433]: You can just check the log output string, but I don't think that it makes any value in the tests.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go [hunk 2]
```diff
@@ -41,6 +42,29 @@ func init() {
 	}
 }
 
+type testLogger struct {
```
[the remark this answers is not in the document]
Comment (danilo-gemoli) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:45]: Because of this solely
```golang
func (log *testLogger) Output() string {
	return log.Logger.Out.(*strings.Builder).String()
}
```

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 1 above)
Comment (danilo-gemoli) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:433]: I want to know how and what the reconciler is going to log because what it has been logging so far wasn't good.
`wantLog` is optional, if you're going to add a new test case and don't want to check any log, just leave it `nil`.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:45]: Why do you need to check the log output on tests? I don't see the point. Also, the whole approach is wrong since you can use the logrus hooks to test this out.

File: pkg/api/multiarchbuildconfig/v1/types.go [hunk 3]
```diff
@@ -66,14 +66,16 @@ func UpdateMultiArchBuildConfig(ctx context.Context, logger *logrus.Entry, clien
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		mabc := &MultiArchBuildConfig{}
 		if err := client.Get(ctx, namespacedName, mabc); err != nil {
+			logger.Error("Failed to get the MultiArchBuildConfig while updating")
```
Comment (droslean) [inline pkg/api/multiarchbuildconfig/v1/types.go:69]: But we are already returning the error. You need to log it on that layer.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 4]
```diff
@@ -235,9 +261,11 @@ func (r *reconciler) handlePushImageWithManifest(ctx context.Context, mabc *v1.M
 			})
 			mabcToMutate.Status.State = v1.FailureState
 		}
+	} else {
```
Comment (droslean) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:264]: No need for the else statement here.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danilo-gemoli) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:45]: I'll assess the hooks and see how the code will look like when using them. To answer your question: https://github.com/openshift/ci-tools/pull/3910#discussion_r1463297795

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 1 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:433]: I don't see any point. It would be better to check the returned error if a test case fail, instead of testing the actual logging.

File: pkg/api/multiarchbuildconfig/v1/types.go (hunk 3 above)
Comment (danilo-gemoli) [reply pkg/api/multiarchbuildconfig/v1/types.go:69]: You mean outside the `RetryOnConflict` every time `UpdateMultiArchBuildConfig` is invoked?

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 1 above)
Comment (danilo-gemoli) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:433]: Well I might revert the tests but we need to improve logs anyway

File: pkg/api/multiarchbuildconfig/v1/types.go (hunk 3 above)
Comment (droslean) [reply pkg/api/multiarchbuildconfig/v1/types.go:69]: This will retry only on conflicts. Technically we already return the error. You can add the log there.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go (hunk 1 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig_test.go:433]: Yea, having better logging is always good, but there is no point in including the logs in the tests.

File: pkg/api/multiarchbuildconfig/v1/types.go (hunk 3 above)
Comment (danilo-gemoli) [reply pkg/api/multiarchbuildconfig/v1/types.go:69]: Something like this?
```diff
if err := v1.UpdateMultiArchBuildConfig(ctx, logger, r.client, ctrlruntimeclient.ObjectKey{Namespace: mabc.Namespace, Name: mabc.Name}, mutateFn); err != nil {
+   logger.Errorf("Failed to update the mabc: %s", err)
    return fmt.Errorf("failed to update the MultiArchBuildConfig %s/%s: %w", mabc.Namespace, mabc.Name, err)
}
```

Comment (droslean) [reply pkg/api/multiarchbuildconfig/v1/types.go:69]: We are returning that error too.

**Comment by danilo-gemoli** [comment]
It seems we have a flaky e2e, @droslean if you're happy with this PR I'm going to retest it again

**Review by droslean — APPROVED** [review]
